Get tests working on ARM64

glevand opened the pull request:
Fixups needed to get the e2e and integration tests working on ARM64.  The two main problems were that the ARM64 kernel optionally supports a 64 KiB page size, and that older ARM64 platforms tend to run slow.

The boltdb sets up its internal memory management based on the system page size, so the system page size must be considered when setting the v3 space quota.  Also, when writing large buffers with etcdctl on 64 KiB page size systems the maximum command line length, generally 128 KiB, needs to be considered.

By running the e2e and integration tests in the integration_pass in series and extending the test timeout values, the integration_pass can be run successfully on some slower systems.

heyitsanthony commented on clientv3/integration/kv_test.go:
```diff
@@ -35,8 +35,8 @@ func TestKVPutError(t *testing.T) {
-		maxReqBytes = 1.5 * 1024 * 1024
-		quota       = int64(maxReqBytes * 1.2)
+		maxReqBytes = 1.5 * 1024 * 1024      // hard coded max in v3_server.go
+		quota       = int64(4 * 1024 * 1024) // the grpc max
```
This test should have nothing to do with grpc limits. What was wrong with `maxReqBytes * 1.2`?

glevand replied:
That was an attempt to resolve this error:

    -- FAIL: TestKVPutError (0.12s)
    kv_test.go:54: expected etcdserver: request is too large, got etcdserver: mvcc: database space exceeded

From what I could tell, ```database space exceeded``` means the quota limit was reached, so I increased the quota to the max, which is 4 MiB (limited by grpc max message size).

heyitsanthony replied:
OK, the problem is that `1.2 * maxReqBytes` doesn't give enough slack for 64kb pages so it's rejecting based on exceeded space instead of max request size. It should be something like `quota = maxReqBytes + 8 * pagesize` so there's enough space for the initial pages.

heyitsanthony commented on clientv3/integration/kv_test.go:
```diff
@@ -59,9 +59,9 @@ func TestKVPutError(t *testing.T) {
+	time.Sleep(1 * time.Second) // give enough time for commit
 
-	_, err = kv.Put(ctx, "foo2", strings.Repeat("a", int(maxReqBytes-50)))
+	_, err = kv.Put(ctx, "foo2", strings.Repeat("a", int(quota-50)))
```
what was wrong with `maxReqBytes - 50`?

glevand replied:
This is testing the over quota alarm, so in the general case, we should be using the quota size to test with.   If we have a quota size of 4MiB, then 1.5 MiB - 50 won't hit the over quota.

heyitsanthony replied:
I think this can stay `maxReqBytes-50` since `2*(maxReqBytes-50) > (maxReqBytes + 8*pagesize)`

heyitsanthony commented on e2e/ctl_v3_alarm_test.go:
```diff
@@ -15,44 +15,62 @@
-		if !strings.Contains(err.Error(), "etcdserver: mvcc: database space exceeded") {
-			cx.t.Fatal(err)
+	// write some small chunks to fill up the database
+	buf := strings.Repeat("b", int(os.Getpagesize()/3))
```
this is a lot less precise than the old code; ideally this would be a single large put that would exceed the quota

glevand replied:
When we run on a system with 64KiB pages, we can't write a buffer large enough to hit the quota limit because we hit the command line limit of etcdctl (128 KiB).  That's why I changed this to fill up the DB with smaller puts until we do hit the quota limit.  What do you suggest?

heyitsanthony replied:
OK, can this put page sized chunks, then? I think that will get around having to defrag+sleep to free up space. The quota limit will see the large request and trigger the alarm. On disarm there should still be enough space to do a small put.

heyitsanthony commented on e2e/ctl_v3_alarm_test.go:
```diff
@@ -15,44 +15,62 @@
-func TestCtlV3Alarm(t *testing.T) { testCtl(t, alarmTest, withQuota(64*1024)) }
+func TestCtlV3Alarm(t *testing.T) {
+	// The boltdb minimum working set is six pages.
+	testCtl(t, alarmTest, withQuota(int64(48*os.Getpagesize())))
```
why 48? I think 12 would be better since that gives a tighter bound on the max quota and the quota can be exceeded with a single Put even with 64k pages.

glevand replied:
I found ```quota=13*pages``` was the smallest value that would not hit the over quota for the first small put.  On my 'fast' system the complete TestCtlV3Alarm test would pass with that.  On the 'slow' system it would still hit the over quota even after the removal + defrag.  Increasing the quota would get the test to pass on the slow system.

For now I'll just set it to ```quota=13*pages```.

I think the problem with the slow sytem may be that it takes a while for the backend/DB to actually free up the space.  If so, how can we assure DB is synced from within the e2e test? 

As above, ```13 * 64 KiB = 832 KiB```.  832 KiB is over the 128 KiB command line length limit of etcdctl, so there's no way to do a single put in the e2e tests.

heyitsanthony commented on e2e/ctl_v3_alarm_test.go:
```diff
@@ -15,44 +15,62 @@
 	}
 
+	// make some space
+	ctlV3Del(cx, []string{"2nd_test", "--prefix"}, 0)
```
this shouldn't be necessary; the large put will trigger the alarm but it won't write the data

glevand replied:
Since my change above loops and fills up the DB with puts, we need to free some space to do this last put.  Without freeing some space we can't put the small buffer.  There may be some combination of quota and buffer sizes that would work on some system, but that would not work in general.

heyitsanthony commented on test:
```diff
@@ -70,14 +78,23 @@ function unit_pass {
+	if [ -n "$SERIAL_TESTS" ]; then
+		echo "Running serial integration tests..."
+		go test -timeout $E2E_TIMEOUT -v -cpu 1,2,4 $@ ${REPO_PATH}/e2e & e2epid="$!"
+		wait $e2epid
```
why have these `wait`s?

glevand replied:
Not needed, I'll drop them.

heyitsanthony commented on test:
```diff
@@ -70,14 +78,23 @@ function unit_pass {
 }
 
 function integration_pass {
-	echo "Running integration tests..."
```
just keep this? I don't think the serial/parallel stuff needs to be written out

glevand replied:
ok

heyitsanthony commented on test:
```diff
@@ -8,6 +8,14 @@
+# Environment variables of interest:
+#
+# E2E_TIMEOUT          - integration_pass e2e test timeout.
+# INTEGRATION_TIMEOUT  - integration_pass integration test timeout.
```
are you using these? I don't think this will work on slower machines since there are a lot of tests that have their own timeout checks like `<-time.After(time.Second)`. The tests should just fail if the machine is too slow; etcd wouldn't work very well if the tests are too slow.

glevand replied:
OK, on working with it some more I agree.

heyitsanthony commented on Documentation/op-guide/maintenance.md:
````diff
@@ -62,7 +62,7 @@ The space quota can be triggered with a loop:
 ```sh
 # fill keyspace
-$ while [ 1 ]; do dd if=/dev/urandom bs=1024 count=1024  | etcdctl put key  || break; done
+$ while [ 1 ]; do dd if=/dev/urandom bs=1024 count=2  | etcdctl put key  || break; done
````
why change this?

glevand replied:
With a small quota value the original code (1 MiB puts) won't fill up the keyspace since it will hit the quota alarm.  2 KiB should always work.

Also, I just noticed this should be ```etcdctl set key```.

heyitsanthony replied:
The point is more to trigger the alarm than to fill it completely. There can still be some space left even with 2k puts.

`etcdctl set` won't work; the quotas only apply to the v3 backend. Probably `ETCDCTL_API=3 etcdctl put key` instead?

heyitsanthony commented:
overall fixes for page size look OK, mostly minor nits in comments. Thanks!

glevand commented:
Rebased to latest.  Addressed all comments so far.  Added a new patch ```Documentation: Improve quota example``` that makes the sample code a little better.

glevand commented:
Fixed gofmt error in kv_test.go.

heyitsanthony commented:
lgtm /cc @xiang90
